sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java

```diff
@@ -48,24 +50,21 @@
 import org.joda.time.Instant;
 
 /**
- * The argument to {@link ParDo} providing the code to use to process
- * elements of the input
- * {@link org.apache.beam.sdk.values.PCollection}.
+ * The argument to {@link ParDo} providing the code to use to process elements of the input {@link
+ * org.apache.beam.sdk.values.PCollection}.
  *
- * <p>See {@link ParDo} for more explanation, examples of use, and
- * discussion of constraints on {@code DoFn}s, including their
- * serializability, lack of access to global shared mutable state,
+ * <p>See {@link ParDo} for more explanation, examples of use, and discussion of constraints on
+ * {@code DoFn}s, including their serializability, lack of access to global shared mutable state,
  * requirements for failure tolerance, and benefits of optimization.
  *
- * <p>{@code DoFn}s can be tested in a particular
- * {@code Pipeline} by running that {@code Pipeline} on sample input
- * and then checking its output.  Unit testing of a {@code DoFn},
- * separately from any {@code ParDo} transform or {@code Pipeline},
- * can be done via the {@link DoFnTester} harness.
+ * <p>{@link DoFn DoFns} can be tested by using {@link TestPipeline}. You can verify their
+ * functional correctness in a local test using the {@code DirectRunner} as well as running
+ * integration tests with your production runner of choice. To create a bundle boundary between
+ * input elements, for testing {@StartBundle} and {@link FinishBundle}, you can use {@link
+ * TestStream} to force quiescence at selected points of your input.
```

Comment:
I'm not sure what "force quiescence" means. It seems that in direct runner indeed 1 `addElements` call generates exactly 1 bundle, at least for the immediately following DoFn, so maybe say "Typically you can generate the input data using Create.of() or other transforms. However, if you need to test behavior of StartBundle and FinishBundle with particular deterministic bundle boundaries, you can use TestStream, where each call to addElements produces 1 bundle", and document that on TestStream#addElements.

Reply:
The semantics of `TestStream` are that all computable consequences of each builder method call must be fully evaluated before going on to the next. So if you `addElements` definitely all of those elements must be fully processed and committed. I admit this is also a potentially undefined notion. But I would say the problem is the spec - certainly the bundle has to be finished and really the sinks should have written everything that is ready based on those elements.

I think each call to `addElements` produces at least one bundle - could be 1 bundle per element. But no element can be left hanging in an unfinished bundle.

Reply:
I'll rephrase to something like what you've said before merging.

Reply:
Yeah it's icky: if you pass TestStream through an identity ParDo and direct runner chose to materialize the results of that, these guarantees all would be bogus. I'm honestly not sure if it's even a good idea to document it anymore. Up to you.

Reply:
Ah, good point. TBH if we can't force bundle boundaries I'm a little sad about this change. I dislike `DoFnTester` a lot but we do need real tests of the IO DoFns. Still, ITs are probably better.

Reply:
Wait, actually that is not true. Materialization does not indicate quiescence. I believe they are required to be processed unless blocked by watermark/trigger progress.

Reply:
I went with your phrasing, cutting off just the last phrase about exactly how it works. Hoping the lack of detail here implies you had better read the docs.